ui/page/components/receive_modal.go

```diff
@@ -0,0 +1,452 @@
+	rm.newAddr.ButtonStyle.TextSize = values.TextSize14
+	rm.newAddr.ButtonStyle.Font.Weight = font.SemiBold
+
+	rm.pageContainer = &widget.List{
+		List: layout.List{
+			Axis:      layout.Vertical,
+			Alignment: layout.Middle,
+		},
+	}
+
+	rm.initWalletSelectors()
+
+	return rm
+}
+
+func (rm *ReceiveModal) OnResume() {
+	rm.ctx, rm.ctxCancel = context.WithCancel(context.TODO())
+
+	rm.sourceWalletSelector.WalletSelected(func(selectedWallet *load.WalletMapping) {
+		rm.sourceAccountSelector.SelectFirstValidAccount(selectedWallet)
+	})
+
+}
+
+func (rm *ReceiveModal) OnDismiss() {
+	rm.ctxCancel()
+}
+
+func (rm *ReceiveModal) Handle() {
+
+	if rm.okBtn.Clicked() || rm.Modal.BackdropClicked(true) {
+		rm.Dismiss()
+	}
+
+	if rm.generateNewAddressButton.Button.Clicked() {
+		rm.generateNewAddress()
+		rm.generateQRForAddress()
+	}
+
+	if rm.more.Button.Clicked() {
+		rm.isNewAddr = !rm.isNewAddr
+		if rm.isInfo {
+			rm.isInfo = false
+		}
+	}
+
+	if rm.newAddr.Clicked() {
+		newAddr, err := rm.generateNewAddress()
+		if err != nil {
+			log.Debug("Error generating new address" + err.Error())
+			return
+		}
+
+		rm.currentAddress = newAddr
+		rm.generateQRForAddress()
+		rm.isNewAddr = false
+	}
+
+	if rm.info.Button.Clicked() {
+		textWithUnit := values.String(values.StrReceive) + " " + string(rm.sourceWalletSelector.selectedWallet.GetAssetType())
+		info := modal.NewCustomModal(rm.Load).
+			Title(textWithUnit).
+			Body(values.String(values.StrReceiveInfo)).
+			SetContentAlignment(layout.NW, layout.W, layout.Center)
+		rm.ParentWindow().ShowModal(info)
+	}
+}
+
+func (rm *ReceiveModal) handleCopyEvent(gtx C) {
+	rm.addressEditor.EditorIconButtonEvent = func() {
+		clipboard.WriteOp{Text: rm.addressEditor.Editor.Text()}.Add(gtx.Ops)
+		rm.Toast.Notify(values.String(values.StrCopied))
+	}
+}
+
+func (rm *ReceiveModal) generateQRForAddress() {
+	rm.addressEditor.Editor.SetText(rm.currentAddress)
+
+	var imgOpt qrcode.ImageOption
+	switch rm.sourceWalletSelector.selectedWallet.Asset.GetAssetType() {
+	case "DCR":
+		imgOpt = qrcode.WithLogoImage(rm.Theme.Icons.DCR)
+	case "BTC":
+		imgOpt = qrcode.WithLogoImage(rm.Theme.Icons.BTC)
+	case "LTC":
+		imgOpt = qrcode.WithLogoImage(rm.Theme.Icons.LTC)
+	}
+
+	qrImage, err := qrcode.New(rm.currentAddress, imgOpt)
+	if err != nil {
+		log.Error("Error generating address qrCode: " + err.Error())
+		return
+	}
+
+	var buffer bytes.Buffer
+	err = qrImage.SaveTo(&buffer)
+	if err != nil {
+		log.Error(err.Error())
+		return
+	}
+
+	decodeImg, _, err := image.Decode(bytes.NewReader(buffer.Bytes()))
+	if err != nil {
+		log.Error(err.Error())
+		return
+	}
+
+	rm.qrImage = &decodeImg
+}
+
+func (rm *ReceiveModal) generateNewAddress() (string, error) {
+
+generateAddress:
+	newAddr, err := rm.sourceWalletSelector.selectedWallet.NextAddress(rm.sourceAccountSelector.selectedAccount.Number)
+	if err != nil {
+		return "", err
+	}
+
+	if newAddr == rm.currentAddress {
+		goto generateAddress
+	}
```

Comment:
Is this possible?

Reply:
yes!, magic of go

Reply:
Yh. I meant to confirm if this case `newAddr == rm.currentAddress` is ever possible, if not we can skip it.

Reply:
oh, my bad, yes there's a bug we discovered a while back, so this was a hack to bypass it, https://github.com/planetdecred/dcrlibwallet/pull/241/files the bug was fixed here but the code of the receive page was never updated, i ported over code from the receive page, i'd update it now, good catch